lib/ts/recipe/multifactorauth/utils.ts

```diff
@@ -0,0 +1,192 @@
+    };
+}
+
+export const isValidFirstFactor = async function (
+    tenantId: string,
+    factorId: string,
+    userContext: UserContext
+): Promise<boolean> {
+    const tenantInfo = await Multitenancy.getTenant(tenantId, userContext);
+    if (tenantInfo === undefined) {
+        throw new Error("tenant not found");
+    }
+    const { status: _, ...tenantConfig } = tenantInfo;
+
+    // we prioritise the firstFactors configured in tenant. If not present, we fallback to the recipe config
+    // Core already validates that the firstFactors are valid as per the logn methods enabled for that tenant,
+    // so we don't need to do additional checks here
+
+    let validFirstFactors =
+        tenantConfig.firstFactors !== undefined
+            ? tenantConfig.firstFactors
+            : MultiFactorAuthRecipe.getInstanceOrThrowError().config.firstFactors;
+
+    if (validFirstFactors === undefined) {
+        // if validFirstFactors is undefined, we can safely assume it to be true because we would then
+        // have other points of failure:
+        // - if login method is disabled in core for the tenant
+        // - if appropriate recipe is not initialized, will result in a 404
+        // In all other cases, we just want to allow all available login methods to be used as first factor
+
+        return true;
+    }
+
+    return validFirstFactors.includes(factorId);
+};
+
+// This function is to reuse a piece of code that is needed in multiple places
+export const getMFARelatedInfoFromSession = async function (
+    input: (
+        | {
+              userId: string;
+              tenantId: string;
+              accessTokenPayload: any;
+          }
+        | {
+              session: SessionContainerInterface;
+          }
+    ) & {
+        assumeEmptyCompletedIfNotFound: boolean;
+        userContext: UserContext;
+    }
+): Promise<
+    | {
+          status: "SESSION_USER_NOT_FOUND_ERROR" | "MFA_CLAIM_VALUE_NOT_FOUND_ERROR" | "TENANT_NOT_FOUND_ERROR";
+      }
+    | {
+          status: "OK";
+          sessionUser: User;
+          factorsSetUpForUser: string[];
+          completedFactors: MFAClaimValue["c"];
+          requiredSecondaryFactorsForUser: string[];
+          requiredSecondaryFactorsForTenant: string[];
+          mfaRequirementsForAuth: MFARequirementList;
+          tenantConfig: TenantConfig;
+      }
+> {
+    let userId: string;
+    let tenantId: string;
+    let accessTokenPayload: any;
+
+    if ("session" in input) {
+        userId = input.session.getUserId();
+        tenantId = input.session.getTenantId();
+        accessTokenPayload = input.session.getAccessTokenPayload();
+    } else {
+        userId = input.userId;
+        tenantId = input.tenantId;
+        accessTokenPayload = input.accessTokenPayload;
+    }
+
+    const sessionUser = await getUser(userId, input.userContext);
+    if (sessionUser === undefined) {
+        return {
+            status: "SESSION_USER_NOT_FOUND_ERROR",
```

Comment:
this can be a misleading error. In case the input is a userId and not a session, then this error is not necessarily about session user not found - correct?

Reply:
We now throw UNAUTHORISED since this function is always called in the context of an session.